Implemented Min/Max Setpoints and Comfort Mode

rkoshak opened the pull request:
This PR implements the dual setpoints for the AC and comfort mode.

Items:
- TempSetpoint is replaced with a MinTempSetpoint used to control the heater and MaxTempSetpoint used to control the AC. MaxTempSetpoint is ignored in EU SystemType. Both has a C and F version as well mirroring the way TempSetpoint works (I don't think they are necessary and will remove them in a later PR). Defaults to use for these setpoints are defined in defaults.js.
- Comfort_Value represents the hysteresis value. This can be any number between 0.5 and 10. The sitemap is set to increment it by .5 when in C mode and 1 when in F mode. The default to use for this Item is defined in defaults.js.
- Comfort_Mode represents the hysteresis mode. If it's "COMFORT" the value defined in defaults.js is used for the hysteresis (0.5 for C and 1 for 2 for F). User can change this by modifying defaults.js. If it's "ECO" the number in Comfort_Value is used for the hysteresis.

All the new Items except for the C and F versions of the setpoints are configured to restoreOnStartup.

Things:
- No new things but there are new Channels on the Sensors and Setpoints Thing to publish and subscribe for the new MaxTempSetpoint, MinTempStepoint, Comfort_Value and Comfort_Mode Items. 

Channel | Subscribes to | Publishes to
-|-|-
MaximumTempSetpoint | hestia/local/cmnd/setmaxtempsetpoint | hestia/local/maxtempsetpoint
MinimumTempSetpoint | hestia/local/cmnd/setmintempsetpoint | hestia/local/mintempsetpoint
Hysteresis | hestia/local/cmnd/hysteresis | hestia/local/hysteresis
Comfort Mode | hestia/local/cmnd/comfortmode | hestia/local/comfortmode

When OH publishes to the topics it does with retained messages.

Rules:
There were surprisingly few Rules that needed to be changed which was nice.

- SystemType Changed: removed references to TempSetpoint, set the MinTempSetpoint and MaxTempSetpoint to the defaults
- Initialization: if NULL or UNDEF, set MinTempSetpoint, MaxTempSetpoint, Comfort_Mode and to the defaults
- Synchronize Temp Proxies: Added MinTempSetpoint, MinTempSetpointF, MinTempSetpoint, MaxTempSetpoint, MaxTempSetpointC, MaxTempSetpointF as triggers so they all keep in sync.
- Added a new Rule, Comfort Mode Changed, which reverts the change if it's anything except "COMFORT" or "ECO"
- Heating/Cooling Check: This is where the main logic resides. Added checking to make sure MinTempSetpoint, MaxTempSetpoint, Comfort_Mode, and Comfort_Value are not NULL or UNDEF before running the Rule, removed reference to TempSetpoint. Rewrote the Action to use the MinTempSetpoint for heating and the MaxTempSetpoint for cooling and to take account the hysteresis. If in COMFORT, the value in defaults.js is used. If in ECO, the value in Comfort_Value is used.
- Sensor or Setpoint Changed: Added MinTempSetpoint, MaxTempSetpoint, Comfort_Value, and Comfort_Mode as triggers. Added a new condition clause to revert and change to Comfort_Value that is outside the 0.5 <= Comfort_Value <= 10 range. If the value is higher 10 is used, if lower 0.5 is used. I may need to revisit this to account for differences in C and F. The Action was modified to call HeatingCheck or CoolingCheck if the Rule was triggered by a change to one of the Comfort Items. If the change was to MinTempSetpoint HeatingCheck is called and if it;s to MaxTempSetpoint CoolingCheck is called. These were the most complex changes made. Removed references to TempSetpoint.
- TempUnit Changed: Removed references to TempSetpoint and added reinitialization of MinTempSetpoint, MaxTempSetpoint, and Comfort_Value to the defaults for the new unit type.

Sitemap:
- In US mode I added the Comfort Mode switch under Temperature and removed the temperature setpoint from that Frame. I added the MinTempSetpoint to the Heating Frame and the MaxTempSetpoint to the Cooling Frame.
- In EU mode I put the Comfort Mode switch under the Heating Frame.
- The Comfort Range (i.e. hysteresis value) I placed under Settings. 

Begins to solve #14 .

rkoshak committed:
Min/max setpoints and hysteresis

Signed-off-by: Richard Koshak <rlkoshak@gmail.com>

rkoshak committed:
Min/max setpoints and hysteresis

rkoshak committed:
Merge branch 'minmax' of github.com:rkoshak/hestia-touch-openhab into minmax

rkoshak committed:
Removed a bunch of stuff that never should have been checked in on the first commit.

Signed-off-by: Richard Koshak <rlkoshak@gmail.com>

jaythomas commented on var/lib/openhab2/jsondb/org.eclipse.smarthome.core.items.Item.json:
```diff
@@ -115,17 +105,14 @@
     }
   },
-  "TempSetpoint": {
+  "MyHumiProxy": {
```
Where are these labels visible in openhab?

rkoshak replied:
In PaperUI you can see all the Items in Configuration -> Items. There is a search function also but it mostly works with Item labels as opposed to Item names (which is what is highlighted).

Personally, I tend to use the REST API Docs, Karaf Console, or look in the JSONDB directly. The problem with this check in is apparently it decided to reorder the entries. 😠 I think this is because I removed some Items. I've only seen it do this much reordering one other time. So many of the changes it is pointing out are false changes caused by the reordering.

I wouldn't have gone down this route of putting everything in JSONDB if the start time savings were not so dramatic.

jaythomas commented on etc/openhab2/automation/lib/hestia/defaults.js:
```diff
@@ -5,14 +5,18 @@ var DEFAULTS = new java.util.HashMap();
 
 // F Defaults
 DEFAULTS.put("F_MIN", 32);             // Minimum F setpoint
 DEFAULTS.put("F_MAX", 100);            // Maximum F setpoint
```
While we're at it, can we change these to something more reasonable, like 60-80? It's a bit of a chore to tap the screen 50 times to get the temperature back when it resets.

rkoshak replied:
I used the defaults that the original code used. These are now in a config file so they can be more easily changed, but an important feature that has been added is that OH now remembers the previous state of these Items and restores them when OH restarts. The only time they should get reset is when you initially start up HestiaPi for the very first time, or when you change the TempUnit or SystemType. In all other cases, you should never have to reset them once adjusted to your preferred values. 

But I'm glad you pointed at this document because it appears to be missing some default values. The default for MinTempSetpoint for C should be 22 and for F it should be 70 and the MaxTempSetpoint for C should be 24 and for F 74. 

I'll get those values checked in right away as you will get errors when starting OH for the very first time without those defaults.

jaythomas commented:
- Any suggestions on an easy way to check this out on a running unit? I'd like to do some live testing on my personal system.
- How does this affect the logic defined in `default.rules`? Will it affect my custom gpio pinout for my 5-wire setup?

rkoshak committed:
Added missing defaults for MinTempSetpoint and MaxTempSetpoint

rkoshak commented:
@gulliverrr is working on a new SD card for testing with. I don't know the status of that but it would be the easiest way to test it out. I wrote a quick post showing how to upgrade and deploy the new code at https://community.hestiapi.com/t/interoperability-with-other-openhab-instance/1586/8 (@gulliverrr, if I recall correctly you found a missing command on that post but I can't find where we had that discussion). The above requires a newer version of openHAB than 2.4.

If possible, I would test on a separate SD card to minimize the disruption to your live system. 

default.rules have been almost completely replaced. It just has a little rule to kick off the JavaScript Initialization Rule. All the Rules are now written in JavaScript and accessible through PaperUI. If you post what you changed I can point you where to make the changes yourself or incorporate the changes for you and post them here with instructions on how to apply them. At a high level, depending on what needs to change, you would disable the default rule(s) and use the REST API to import the new Rule (a curl command). 

I wrote up a general guide for the best way to customize the Rules at https://community.hestiapi.com/t/how-to-customize-a-rule/1644. Following this approach it is easier to preserve customization during future upgrades (tutorial for that is to be written). 

I'm also working on some activity diagrams to help explain how the individual Rules work together. Here are those I've done so far. Note, they do not reflect these changes yet (i.e. MinTempSetpoint, MaxTempSetpoint, Comfort).

![image](https://user-images.githubusercontent.com/6385096/83272067-f289de80-a187-11ea-9c8a-ae4e805f6b22.png)

![image](https://user-images.githubusercontent.com/6385096/83272115-033a5480-a188-11ea-997f-0a86ea43ec2d.png)

![image](https://user-images.githubusercontent.com/6385096/83272186-19e0ab80-a188-11ea-9e94-2ade43030527.png)

![image](https://user-images.githubusercontent.com/6385096/83272289-32e95c80-a188-11ea-9995-e4f9ecd546b6.png)

![image](https://user-images.githubusercontent.com/6385096/83272339-41d00f00-a188-11ea-806d-8b1932a5c752.png)

![image](https://user-images.githubusercontent.com/6385096/83272477-6cba6300-a188-11ea-8aa4-bffd7b4aea4b.png)

![image](https://user-images.githubusercontent.com/6385096/83272550-82c82380-a188-11ea-84ef-8c1555a26ee8.png)

![image](https://user-images.githubusercontent.com/6385096/83272605-94113000-a188-11ea-97cf-cecdc81a43fc.png)

![image](https://user-images.githubusercontent.com/6385096/83272648-a2f7e280-a188-11ea-9cf0-7cc7180c8635.png)


NOTE: These drawings were made using draw.io and the source for them will be checked in somewhere so they can be modified as needed moving forward.

jaythomas commented:
Nice work on diagram and documenting. I would suggest checking the diagrams in this git repo so changes can be tracked with the code.

> @gulliverrr is working on a new SD card for testing with.

So should I wait on that image before testing?

Regarding the changes I've made to the `default.rules`, I've only modified the pins that are being switched on and off as you can see the pin I added to energize the reversing valve in my heat pump for cooling:

```
rule "CoolingPin changed"
when
    Item CoolingPin changed
then
    switch(CoolingPin.state) {
      case ON: {
        Pin23.sendCommand(ON)
        Pin12.sendCommand(ON)
        //CoolingPinTopic.sendCommand(ON)
        logInfo("Default","CoolingPin set to ON.")
      }
      case OFF: {
        Pin12.sendCommand(OFF)
        Pin23.sendCommand(OFF)
        //CoolingPinTopic.sendCommand(OFF)
        logInfo("Default","CoolingPin set to OFF.")
      }
    }
end
```

If that can be modified without needing to ssh into the unit that's great, but if this has moved to another script file just let me know where I need to change.

gulliverrr commented:
> So should I wait on that image before testing?

@jaythomas I haven't checked the PR yet but the new image should not affect OH files/configuration that much. Some HestiaPi scripts (under /home/pi/scripts) will be modified but shouldn't stop you from testing UI stuff.

rkoshak commented:
> Regarding the changes I've made to the default.rules, I've only modified the pins that are being switched on and off as you can see the pin I added to energize the reversing valve in my heat pump for cooling:

OK, that's not too bad. Are you running in US mode and turning on the CoolingPin (Pin23) and the HeatingPin (Pin12) at the same time or are you running in EU mode and turning on the HeatingPin (Pin23) and the HotWaterPin (Pin12) at the same time? 

The mode will determine the best approach to take. In US mode, what about the Fan (Pin18)? If in EU mode, I think it's a bit more straight forward.

But if this Rule exists at all, I imagine that you are in US mode since as, if I recall correctly, there is no "CoolingPin changed" Rule in the EU rules file. So if I assume US mode I need to understand how the Fan works in your system. Do you turn it on only for heating? Which pins control the heating since you are using the same pin here? Or is it the case that during heating you turn on Pin12 (Heating) and Pin18 (Fan) and for cooling you turn on Pin12 (Heating) and Pin23 (Cooling)?

So I think I need to know more to tell you what to change. If I make assumptions 

Changed to make to the "Cooling Control" Rule (this Rule get's called when the "Heating/Cooling Check" Rule determines the AC needs to be turned ON or OFF:

- Modify the "The command is different from the current cooling and fan states" condition to remove the checking of the Fan states

```javascript
(items["CoolingPin"] != command);
```

- Delete the "Prevent the Heating and Cooling from being ON at the same time" condition entirely

- In the Action, remove all references to the Fan and add lines to sendCommand ON/OFF to HeatingPin in addition to CoolingPin. The timer that makes sure the Fan turns on before the AC turns on can be removed as well

```javascript
ar OPENHAB_CONF = Java.type('java.lang.System').getenv('OPENHAB_CONF');
load(OPENHAB_CONF + '/automation/lib/hestia/utils.js');

logInfo("cooling", "Turning " + command + " the cooling: curr temp = " + items["MyTemp"] +
                 " setpoint = " + items["MaxTempSetpoint"] +
                 " mode = " + items["CoolingMode"]);

if(command == ON){
  commandIfDifferent("MainSwitch", ON);
  commandIfDifferent("CoolingPin", ON);
  commandIfDifferent("HeatingPin", ON);
  events.sendCommand("CoolingMode", items["CoolingMode"]); // update the LCD
}
else {
  commandIfDifferent("CoolingPin", OFF);
  commandIfDifferent("HeatingPin", OFF);
}
```

And then under the Rule "Map Pin to GPIO" (this Rule takes the Pin Items (e.g. HeatingPin, CoolingPin) and sends the command to the appropriate GPIO Pin Item (e.g. Pin23) based on the mapping defined in defaults.js:

- The "Fan Protection" condition needs to be modified to only make sure the Fan cannot be turned OFF when the Heating is ON but the Cooling is OFF. Obviously this is only required if the Fan does indeed need to come on when Heating. If not you can probably eliminate this condition entirely. 

```javascript
var OPENHAB_CONF = Java.type('java.lang.System').getenv('OPENHAB_CONF');
load(OPENHAB_CONF + '/automation/lib/hestia/utils.js');

var ok = true
if(event.itemName == "FanPin"){
  logInfo("pins", "FanPin commanded to " + newState);
  if(items["SystemType"] == "US" &&
     (items["HeatingPin"] == ON && items["CoolingPin"] == OFF) &&
     newState == OFF){
    logWarn("pins", "Cannot turn off the fan when heating! Restoring fan states!");
    commandIfDifferent("FanMode", "AUTO");
    commandIfDifferent("FanCtrl", "OFF");
    events.sendCommand("FanPin", ON);
    ok = false;
  }
}
(ok);
```


> If that can be modified without needing to ssh into the unit that's great, but if this has moved to another script file just let me know where I need to change.

All Rules can be edited through PaperUI now. But the overall way the Rules work is significantly different from the way the old Rules worked, avoiding lots of duplicated code and adding in a lot more checks. My second link in the previous post should give you the orientation you need to modify the Rules using just the browser and hopefully the activity diagrams above show how the events move from one Rule to the next (the rounded corner rectangles are each individual Rules).  One of my goals with moving everything to JSONDB, beyond the startup improvements, was to make as much as is possible viewable and modifiable through PaperUI so users almost never need to ssh to the machine to modify the behaviors. Certain goals like setting up email notifications using only the browser interface are only possible with this as well. 

As described in the link above, disable and clone the Rules before modifying them so you have easy reference back to the original. Also, any change you make will automatically create a backup in /var/openhab2/jsondb/backup, but it's much easier to just review the disabled Rule in PaperUI).

jaythomas commented:
Thanks for the info, I think I can figure it out. What's the easiest way to deploy this code to a unit that I've imaged?

gulliverrr commented:
@jaythomas should I wait for your commit on the UI to merge to main?

jaythomas commented:
@gulliverrr I haven't tested it to know if its stable enough to merge into master, but that's up to you.

gulliverrr commented:
Sure, I didn't mean that. Actually it will make no difference to have it checked in before your UI commit so I will go ahead and continue testing it and merge it.

rkoshak commented:
> What's the easiest way to deploy this code to a unit that I've imaged?

Follow the steps at https://community.hestiapi.com/t/interoperability-with-other-openhab-instance/1586/8.

jaythomas commented:
Thanks, didn't see the thread originally. I'm going to fire this up on a live unit now and test it throughout the evening.

gulliverrr commented:
> @gulliverrr, if I recall correctly you found a missing command on that post but I can't find where we had that discussion

https://community.hestiapi.com/t/interoperability-with-other-openhab-instance/1586/16

It is only for changing the default (F) temp unit so one could get away without it... I have added this to my new image

> NOTE: These drawings were made using draw.io and the source for them will be checked in somewhere so they can be modified as needed moving forward.

I would suggest the [docs](https://github.com/HestiaPi/hestia-touch-openhab/tree/ONE/docs) folder. Definitely worth checking them in in code and png format so that I link to them from the [wiki](https://github.com/HestiaPi/hestia-touch-openhab/wiki).

rkoshak commented:
> I would suggest the docs folder. Definitely worth checking them in in code and png format so that I link to them from the wiki.

Will do. I need to complete them and update them for the latest changes first. Expect a PR sometime this week.
